### ISSUE TITLE
Fluentd pod failed to become ready on master

### ISSUE THREAD

**Issue by MrHohn** [kubernetes/kubernetes#54736]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind bug

**What happened**:
Fluentd pod failed to become ready on master after upgrade from 1.8->HEAD. 

Logs from https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1043:
```
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:260
Oct 27 16:30:05.440: Error waiting for all pods to be running and ready: 1 / 29 pods in namespace "kube-system" are NOT in RUNNING and READY state in 10m0s
POD                      NODE                 PHASE   GRACE CONDITIONS
fluentd-gcp-v2.0.9-4pp8p bootstrap-e2e-master Running       [{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-27 16:02:00 +0000 UTC Reason: Message:} {Type:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-27 16:13:10 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: [fluentd-gcp prometheus-to-sd-exporter]} {Type:PodScheduled Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-27 16:02:03 +0000 UTC Reason: Message:}]

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/e2e.go:198
```

Multiple CIs are hitting this:
- https://k8s-testgrid.appspot.com/sig-release-master-upgrade#gce-1.8-master-upgrade-master
- https://k8s-testgrid.appspot.com/sig-network#gci-gce-latest-upgrade-kube-proxy-ds
- https://k8s-testgrid.appspot.com/sig-network#gci-gce-latest-downgrade-kube-proxy-ds

**Anything else we need to know?**:
From the timeline, possibly caused by the fluentd hostnetwork change:
- https://github.com/kubernetes/kubernetes/pull/54395
- https://github.com/kubernetes/kubernetes/pull/54694

@crassirostris @loburm

**Comment by MrHohn** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-alpha-features/17205 has the same symptom as well.

/sig instrumentation

**Comment by crassirostris** [comment]
```
Reason:ContainersNotReady Message:containers with unready status: [prometheus-to-sd-exporter]
```

@loburm Could you please take a look?

**Comment by crassirostris** [comment]
@MrHohn It doesn't seem like an upgrade issue actually. It happens exclusively on master and I see it happening just after regular cluster startup. In the prom-to-sd logs:

```
F1028 21:36:36.001045       1 main.go:81] Failed to get GCE config: error while getting project id: Get http://169.254.169.254/computeMetadata/v1/project/project-id: dial tcp 169.254.169.254:80: i/o timeout
```

Why is metadata server unaccessible from the master? @ihmccreery Can it be related to the metadata-proxy?

**Comment by MrHohn** [comment]
> @mrhohn It doesn't seem like an upgrade issue actually. It happens exclusively on master and I see it happening just after regular cluster startup

Yeah I agree, edited the title. Sorry for the confusion.

**Comment by ihmccreery** [comment]
This is definitely metadata proxy related.

Are you still seeing this?  I introduced a bug in https://github.com/kubernetes/kubernetes/pull/54150 that I fixed in https://github.com/kubernetes/kubernetes/pull/54725.

**Comment by crassirostris** [comment]
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1078

Same issue today. Again on master

**Comment by ihmccreery** [comment]
Okay.  My guess is that the metadata proxy isn't getting scheduled on the master for some reason.  Looking ...

**Comment by ihmccreery** [comment]
Hm.  The master's [kubelet log](https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-gce-new-master-upgrade-master/1078/artifacts/bootstrap-e2e-master/kubelet.log) shows that it's got the right node label and that a proxy is being scheduled:

```
I1031 17:06:08.380024    1448 flags.go:52] FLAG: --node-labels="beta.kubernetes.io/fluentd-ds-ready=true,beta.kubernetes.io/metadata-proxy-ready=true"
[...]
I1031 17:07:04.446869    1448 config.go:383] Receiving a new pod "metadata-proxy-v0.1-cchvf_kube-system(eb1f4bb8-be5d-11e7-bb2d-42010a800002)"
[...]
I1031 17:07:04.657141    1448 status_manager.go:472Å Status for pod "metadata-proxy-v0.1-cchvf_kube-system(eb1f4bb8-be5d-11e7-bb2d-42010a800002)" updated successfully: (1, æPhase:Pending Conditions:ÆæType:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-31 17:07:04 +0000 UTC Reason: Message:å æType:Ready Status:False LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-31 17:07:04 +0000 UTC Reason:ContainersNotReady Message:containers with unready status: Æmetadata-proxyÅåÅ Message: Reason: HostIP:10.128.0.2 PodIP:10.128.0.2 StartTime:2017-10-31 17:07:04 +0000 UTC InitContainerStatuses:ÆÅ ContainerStatuses:ÆæName:metadata-proxy State:æWaiting:&ContainerStateWaitingæReason:ContainerCreating,Message:,å Running:nil Terminated:nilå LastTerminationState:æWaiting:nil Running:nil Terminated:nilå Ready:false RestartCount:0 Image:gcr.io/google-containers/metadata-proxy:0.1.3 ImageID: ContainerID:åÅ QOSClass:Guaranteedå)
[...]
I1031 17:07:13.682558    1448 status_manager.go:472] Status for pod "metadata-proxy-v0.1-cchvf_kube-system(eb1f4bb8-be5d-11e7-bb2d-42010a800002)" updated successfully: (2, {Phase:Running Conditions:[{Type:Initialized Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-31 17:07:04 +0000 UTC Reason: Message:} {Type:Ready Status:True LastProbeTime:0001-01-01 00:00:00 +0000 UTC LastTransitionTime:2017-10-31 17:07:13 +0000 UTC Reason: Message:}] Message: Reason: HostIP:10.128.0.2 PodIP:10.128.0.2 StartTime:2017-10-31 17:07:04 +0000 UTC InitContainerStatuses:[] ContainerStatuses:[{Name:metadata-proxy State:{Waiting:nil Running:&ContainerStateRunning{StartedAt:2017-10-31 17:07:13 +0000 UTC,} Terminated:nil} LastTerminationState:{Waiting:nil Running:nil Terminated:nil} Ready:true RestartCount:0 Image:gcr.io/google-containers/metadata-proxy:0.1.3 ImageID:docker-pullable://gcr.io/google-containers/metadata-proxy@sha256:279ed21137afcaba2d1e546cc859d3a6c8061a5daeb2f9e854a4825d2cb1fa6f ContainerID:docker://14faf4e0d44bfc58ff628698608d9178a55c6e835d32eb7ff038817362deef48}] QOSClass:Guaranteed})
```

**Comment by ihmccreery** [comment]
Okay, I'm seeing the same thing on a newly-created HEAD cluster.  Metadata-proxy is running and serving requests ... my next guess is that the iptables rules are messed up.

**Comment by ihmccreery** [comment]
The iptables rules are fine:

```
e2e-test-ihmccreery-master ihmccreery # iptables -S
-P INPUT DROP
-P FORWARD DROP
-P OUTPUT DROP
-N DOCKER
-N DOCKER-ISOLATION
-N KUBE-FIREWALL
-N KUBE-METADATA-SERVER
-A INPUT -j KUBE-FIREWALL
-A INPUT -m state --state RELATED,ESTABLISHED -j ACCEPT
-A INPUT -i lo -j ACCEPT
-A INPUT -p icmp -j ACCEPT
-A INPUT -p tcp -m tcp --dport 22 -j ACCEPT
-A INPUT -p tcp -j ACCEPT
-A INPUT -p udp -j ACCEPT
-A INPUT -p icmp -j ACCEPT
-A FORWARD -d 169.254.169.254/32 -p tcp -m tcp --dport 80 -j KUBE-METADATA-SERVER
-A FORWARD -j DOCKER-ISOLATION
-A FORWARD -o docker0 -j DOCKER
-A FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A FORWARD -i docker0 ! -o docker0 -j ACCEPT
-A FORWARD -i docker0 -o docker0 -j ACCEPT
-A FORWARD -p tcp -j ACCEPT
-A FORWARD -p udp -j ACCEPT
-A FORWARD -p icmp -j ACCEPT
-A OUTPUT -j KUBE-FIREWALL
-A OUTPUT -m state --state NEW,RELATED,ESTABLISHED -j ACCEPT
-A OUTPUT -o lo -j ACCEPT
-A DOCKER-ISOLATION -j RETURN
-A KUBE-FIREWALL -m comment --comment "kubernetes firewall for dropping marked packets" -m mark --mark 0x8000/0x8000 -j DROP
-A KUBE-METADATA-SERVER -j ACCEPT
-A KUBE-METADATA-SERVER -j DROP
```

**Comment by ihmccreery** [comment]
cc @dnardo

**Comment by Random-Liu** [comment]
Note that we moved fluentd out of host network recently. https://github.com/kubernetes/kubernetes/pull/54395

**Comment by ihmccreery** [comment]
Thanks, yup. We're trying to root cause; unclear if that's what caused it, or a change in the metadata proxy.

**Comment by Random-Liu** [comment]
Thanks @dnardo for debugging this!
@dnardo found that `net.ipv4.conf.all.route_localnet` on master node is 0, which should be 1.

I set it to 1, and it does fix the issue.

**Comment by Random-Liu** [comment]
I've included the fix in https://github.com/kubernetes/kubernetes/pull/54838.

**Comment by crassirostris** [comment]
@ihmccreery @dnardo @Random-Liu Thanks a lot for debugging!

@Random-Liu Could you please add `fixes #54736` to the PR description?

Also, what about previous branches? Is it going to work on 1.7 and 1.8 branches, or we need to propagate your fix there?
